### PR TITLE
Update FastDelaunay.h

### PR DESCRIPTION
**Pull request by FlowerOfSorrow**
1. Fixed memory leak.
2. Returns the indices instead.

I have another version with only memory leak fix, which seems to be around 20% faster than this one.

In case I messed up, you can find both of the stress testing code below.

https://pastebin.com/kiF4uTkQ (no memory leak)
https://pastebin.com/5Lg9X4Am (no memory leak, returns indices, but about 20% slower than the above)

I think the way I free the memory in both versions is kinda inefficient, but I can't think of a better way for now.

### CODE REVIEW

**Review by simonlindholm — COMMENTED** [review]
Sorry for taking a while to get to this. I feel kinda reluctant to take a 20% perf hit, the Delaunay triangulation is slow enough as it is, and the index-returning version looks to add a bit of code size too. Seems simpler to let the user lookup point -> index in a map afterwards.

It's similarly unfortunate that releasing memory also decreases perf (10% according to discussion in AC Discord). So here's an idea: we could hand-roll our own memory allocator (it's not hard to beat malloc/free for fixed-sized allocations). Quad already contains Quad pointers which we can use as a singly linked free list, and we always alloc/free four circularly linked Quad's at a time, which simplifies the (de)allocation code. I tried this here: https://github.com/kth-competitive-programming/kactl/compare/master...simonlindholm:delaunay-manual-allocator. Appears to reduce runtime on the -DTEST_PERF stress-test from ~0.41s to ~0.27s while adding only ~1 line, so it seems like a nice win! I don't know how many Quad's we need to preallocate though, `sz(pts) * 16 + 60` was a pure guess. (sz(pts) * 8 + 60 crashes.) Insight there would be welcome.

**Comment by simonlindholm** [comment]
It's also possible to use `new` instead of pre-allocating; this wins a line but loses a bit of performance (~0.32s). See the newly added commit in the compare view linked above.

**Comment by FlowerOfSorrow** [comment]
> It's similarly unfortunate that releasing memory also decreases perf (10% according to discussion in AC Discord). So here's an idea: we could hand-roll our own memory allocator (it's not hard to beat malloc/free for fixed-sized allocations). Quad already contains Quad pointers which we can use as a singly linked free list, and we always alloc/free four circularly linked Quad's at a time, which simplifies the (de)allocation code. I tried this here: [master...simonlindholm:delaunay-manual-allocator](https://github.com/kth-competitive-programming/kactl/compare/master...simonlindholm:delaunay-manual-allocator). Appears to reduce runtime on the -DTEST_PERF stress-test from ~0.41s to ~0.27s while adding only ~1 line, so it seems like a nice win!

Wow that sounds really nice.

> I don't know how many Quad's we need to preallocate though, `sz(pts) * 16 + 60` was a pure guess. (sz(pts) * 8 + 60 crashes.) Insight there would be welcome.

Worst case in terms of memory is when their convex hull is a triangle, which gives 3 * sz(ptz) - 6 edges. We get an instance of Quad for each pairing of an edge and one of its 4 orientation, which gives 12 * sz(ptz) - 24 Quads at worst.

File: content/geometry/FastDelaunay.h [hunk 1]
```diff
@@ -91,16 +92,24 @@ pair<Q,Q> rec(const vector<P>& s) {
...
+	while (pts[e->o->F()].cross(pts[e->F()], pts[e->p]) < 0) e = e->o;
+#define ADD { Q c = e; do { c->mark = 1; temp.push_back(c->p); \
+	q.push_back(c->r()); rem.push_back(c); c = c->next(); } while (c != e); }
+	ADD; temp.clear();
 	while (qi < sz(q)) if (!(e = q[qi++])->mark) ADD;
-	return pts;
+	assert((int)temp.size() % 3 == 0);
+	for(auto e: rem) delete e->rot, delete e;
+	vector<array<int, 3>> res;
+	for(int i = 0; i < sz(temp); i += 3) if(pts[temp[i]].cross(pts[temp[i + 1]], pts[temp[i + 2]]) > 0)
```
[the remark this answers is not in the document]
Comment (FlowerOfSorrow) [reply content/geometry/FastDelaunay.h:112]: Actually that if-statement always evaluate to true regardless of the input I think. Not sure why did I put it there.

**Comment by simonlindholm** [comment]
Hrm, 12 * sz(pts) appears to crash, actually... I think I'll go with the `new` based solution, which is a bit shorter and less magical. I opened #196 for that; will close this one. Thanks for looking into this, and all the insights on it!

**Pull request closed without merging**